[PATCH API-NEXT v5] example: ipsec_offload: add ipsec_offload example

lumag opened the pull request:
This is based on old IPsec example by @NikhilA-Linaro , just updated to new packet-API.

[DES: adapted example to updated API.]

Signed-off-by: Nikhil Agarwal <nikhil.agarwal@linaro.org>
Signed-off-by: Dmitry Eremin-Solenikov <dmitry.ereminsolenikov@linaro.org>

muvarov commented:
for v2: this can be fixed:
WARNING: else is not generally useful after a break or return
#435: FILE: example/ipsec_offload/odp_ipsec_offload.c:343:
+		return PKT_POSTED;
+	} else {

lumag commented:
Current code is better, so I don't plan to fix it

Bill-Fischofer-Linaro commented on example/ipsec_offload/odp_ipsec_offload.c:
```diff
@@ -0,0 +1,855 @@
+			return PKT_DROP;
+		}
+	}
+}
```
Checkpatch doesn't like this and the `goto` is a bit ugly. Might a restructure along the following lines be simpler and cleaner?
```
if (!flow) {
        /*Check into Routing table*/
        fwd_entry = find_fwd_db_entry(dip);
     
        if (!fwd_entry) {
			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
			odp_packet_free(pkt);
 	                return PKT_DROP;
        }
        ... flow setup processing, fall though without needing a goto.
}

...flow processing goes here
```

Bill-Fischofer-Linaro commented on example/ipsec_offload/odp_ipsec_offload_fwd_db.c:
```diff
@@ -0,0 +1,221 @@
+/** Global pointer to fwd db */
+fwd_db_t *fwd_db;
+
+void odp_init_routing_table(void)
```
Avoid `odp_` prefix use for non-APIs.

Bill-Fischofer-Linaro commented on example/ipsec_offload/odp_ipsec_offload_fwd_db.h:
```diff
@@ -0,0 +1,201 @@
+ *
+ * @return Matching flow entry
+ */
+static inline odp_flow_entry_t *odp_route_flow_lookup_in_bucket(uint32_t sip,
```
Drop `odp_` prefix

Bill-Fischofer-Linaro commented on example/ipsec_offload/odp_ipsec_offload_fwd_db.h:
```diff
@@ -0,0 +1,201 @@
+ * @param bucket	Target Hash Bucket
+ *
+ */
+static inline void odp_route_flow_insert_in_bucket(odp_flow_entry_t *flow,
```
Drop `odp_` prefix.

lumag commented:
@Bill-Fischofer-Linaro I have dropped `odp_` prefix from function names and reworked this `if (flow)` part. Regarding files names. All examples use `odp_` prefix for file names, so I'm not changing that.

 @bala-manoharan @NikhilA-Linaro @psavol I feel uneasy about using 'both' to specify ESP packet with  authentication and just 'esp' to specify encrypted ESP packet w/o auth. I might want to change that in next revision.

muvarov commented on example/ipsec_offload/Makefile.am:
```diff
@@ -0,0 +1,19 @@
+include $(top_srcdir)/example/Makefile.inc
+
+bin_PROGRAMS = odp_ipsec_offload$(EXEEXT)
```
EXEEXT needs to be removed.

muvarov commented on example/ipsec_offload/Makefile.am:
```diff
@@ -0,0 +1,19 @@
+			 odp_ipsec_offload_sa_db.c \
+			 odp_ipsec_offload_sp_db.c \
+			 odp_ipsec_offload_fwd_db.c \
+			 odp_ipsec_offload_cache.c
```
EXTRA_DIST for bash scripts which should be named correctly.

muvarov commented on example/ipsec_offload/Makefile.am:
```diff
@@ -0,0 +1,19 @@
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_sp_db.h \
+		  $(top_srcdir)/example/example_debug.h
+
+dist_odp_ipsec_offload_SOURCES = odp_ipsec_offload.c \
```
just _SOURCES like all other example/*/Makefile.am

muvarov commented on example/ipsec_offload/Makefile.am:
```diff
@@ -0,0 +1,19 @@
+
+bin_PROGRAMS = odp_ipsec_offload$(EXEEXT)
+odp_ipsec_offload_LDFLAGS = $(AM_LDFLAGS) -static
+odp_ipsec_offload_CFLAGS = $(AM_CFLAGS) -I${top_srcdir}/example
```
no need for CFLAGS/LDFLAGS

muvarov commented on example/ipsec_offload/odp_ipsec_offload_cache.c:
```diff
@@ -0,0 +1,146 @@
+	shm = odp_shm_reserve("shm_ipsec_cache",
+			      sizeof(ipsec_cache_t),
+			      ODP_CACHE_LINE_SIZE,
+			      0);
```
shm != ODP_SHM_INVALID

muvarov commented on example/ipsec_offload/odp_ipsec_offload_fwd_db.c:
```diff
@@ -0,0 +1,221 @@
+	/*Reserve memory for Routing hash table*/
+	hash_shm = odp_shm_reserve("route_table",
+				   sizeof(flow_bucket_t) * bucket_count,
+				   ODP_CACHE_LINE_SIZE, 0);
```
!= ODP_SHM_INVALID

muvarov commented on example/ipsec_offload/odp_ipsec_offload_fwd_db.h:
```diff
@@ -0,0 +1,201 @@
+ *
+ * @param entries number of entries
+ *
+ * @return 0 if successful else -1
```
why this file has bunch of empty lines like here? On  other files code is more compact. Let's keep the same code style.

muvarov commented on example/ipsec_offload/odp_ipsec_offload_misc.h:
```diff
@@ -0,0 +1,366 @@
+#include <odp/helper/ipsec.h>
+
+#ifndef TRUE
+#define TRUE  1
```
looks a little bit ugly, but it's ok.

lumag replied:
I did not want to be too intrusive into this example.

muvarov commented:
@lumag git commit has changelog inside it. This needs to be cleaned up.

lumag commented:
@muvarov updated

lumag commented:
@muvarov It's not quite a changelog. It is a typical notion of changes done by person who applied the commit, rather than by original author.

muvarov commented:
Merged.
